modules/misc/src/materials/ADDensity.C

```diff
@@ -0,0 +1,137 @@
+    params.addCoupledVar("disp_z", "The z displacement");
+    params.addCoupledVar(
+        "displacements",
+        "The displacements appropriate for the simulation geometry and coordinate system");
+    params.addRequiredParam<Real>("density", "Initial density"););
+
+template <ComputeStage compute_stage>
+ADDensity<compute_stage>::ADDensity(const InputParameters & parameters)
+  : ADMaterial<compute_stage>(parameters),
+    _is_coupled(true),
+    _disp_r(isCoupled("displacements")
+                ? adCoupledValue("displacements", 0)
+                : (isCoupled("disp_r") ? adCoupledValue("disp_r") : adZero())),
+    _initial_density(adGetParam<Real>("density")),
+    _density(adDeclareADProperty<Real>("density"))
+{
+  const bool is_r_coupled = isCoupled("disp_r");
+  const bool is_x_coupled = isCoupled("disp_x");
+  const bool is_y_coupled = isCoupled("disp_y");
+  const bool is_z_coupled = isCoupled("disp_z");
+
+  // new parameter scheme
+  if (isCoupled("displacements"))
+  {
+    // get coordinate system
+    _coord_system = getBlockCoordSystem();
+
+    // get coupled gradients
+    const unsigned int ndisp = coupledComponents("displacements");
+    _grad_disp.resize(ndisp);
+    for (unsigned int i = 0; i < ndisp; ++i)
+      _grad_disp[i] = &adCoupledGradient("displacements", i);
+
+    // fill remaining components with zero
+    _grad_disp.resize(3, &adGradZero());
+  }
+
+  // old deprecated parameters
+  else if (is_x_coupled || is_r_coupled)
+  {
+    // guess(!) coordinate system
+    if (is_r_coupled)
+    {
+      if (is_z_coupled)
+        _coord_system = Moose::COORD_RZ;
+      else
+        _coord_system = Moose::COORD_RSPHERICAL;
+    }
+    else
+      _coord_system = Moose::COORD_XYZ;
+
+    // couple gradients
+    _grad_disp = {is_x_coupled ? &adCoupledGradient("disp_x")
+                               : (is_r_coupled ? &adCoupledGradient("disp_r") : &adGradZero()),
+                  is_y_coupled ? &adCoupledGradient("disp_y")
+                               : (is_z_coupled ? &adCoupledGradient("disp_z") : &adGradZero()),
+                  _coord_system != Moose::COORD_RZ && is_z_coupled ? &adCoupledGradient("disp_z")
+                                                                   : &adGradZero()};
+  }
+
+  // no coupling
+  else
+  {
+    _is_coupled = false;
+    // TODO: We should deprecate this case and have the user use a GenericConstantMaterial for this
+  }
+}
+
+template <ComputeStage compute_stage>
+void
+ADDensity<compute_stage>::initQpStatefulProperties()
+{
+  _density[_qp] = _initial_density;
+}
+
+template <ComputeStage compute_stage>
+void
+ADDensity<compute_stage>::computeQpProperties()
+{
+  if (_is_coupled)
+  {
+    // rho * V = rho0 * V0
+    // rho = rho0 * V0 / V
+    // rho = rho0 / det(F)
+    // rho = rho0 / det(grad(u) + 1)
+
+    const auto Axx = (*_grad_disp[0])[_qp](0) + 1.0;
+    const auto Axy = (*_grad_disp[0])[_qp](1);
```

Comment:
Let's use `const auto &` to prevent construction of another dual number. Same goes for Axz, Ayx, Ayz, Azx, Azy